wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Create/Logic/GroupConversationCreationCoordinator.swift

```diff
@@ -0,0 +1,218 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireDataModel
+
+protocol GroupConversationCreator {
+    func insertGroupConversation(
+        moc: NSManagedObjectContext,
+        participants: [ZMUser],
+        name: String?,
+        team: Team?,
+        allowGuests: Bool,
+        allowServices: Bool,
+        readReceipts: Bool,
+        messageProtocol: MessageProtocol) -> ZMConversation?
+}
+
+class DefaultGroupConversationCreator: GroupConversationCreator {
+    func insertGroupConversation(
+        moc: NSManagedObjectContext,
+        participants: [ZMUser],
+        name: String?,
+        team: Team?,
+        allowGuests: Bool,
+        allowServices: Bool,
+        readReceipts: Bool,
+        messageProtocol: MessageProtocol) -> ZMConversation? {
+            return ZMConversation.insertGroupConversation(
+                moc: moc,
+                participants: participants,
+                name: name,
+                team: team,
+                allowGuests: allowGuests,
+                allowServices: allowServices,
+                readReceipts: readReceipts,
+                messageProtocol: messageProtocol
+            )
+        }
+}
+
+class GroupConversationCreationCoordinator {
+    private enum State {
+        case ready
+        case initializing
+        case initialized(eventHandler: (GroupConversationCreationEvent) -> Void)
+        case beginingToCreateConversation(eventHandler: (GroupConversationCreationEvent) -> Void)
+        case creatingConversation(conversation: ZMConversation, eventHandler: (GroupConversationCreationEvent) -> Void)
+        case finalizing
+    }
+    private var state: State = .ready
+
+    private var conversation: ZMConversation? {
+        guard case .creatingConversation(let conversation, _) = state else { return nil }
+        return conversation
+    }
+
+    private var eventHandler: ((GroupConversationCreationEvent) -> Void)? {
+        switch state {
+        case .creatingConversation(_, let completionHandler), .initialized(let completionHandler), .beginingToCreateConversation(let completionHandler):
+            return completionHandler
+        case .ready, .finalizing, .initializing:
+            return nil
+        }
+    }
+    
+    private var creator: GroupConversationCreator
+    
+    init(creator: GroupConversationCreator) {
```

Comment:
if this is just for mocking, you could provide a default value for anything that isn't tests
```suggestion
    init(creator: GroupConversationCreator = GroupConversationCreator()) {
```